File: 695-max-area-of-island/max-area-of-island.cpp
class Solution {
public:

    int n,m;
    void dfs(int row,int col,vector<vector<int>>& grid,vector<vector<int>>& vis,int &count)
    {
        if(row<0 or row>=n or col<0 or col>=m or grid[row][col]==0 or vis[row][col]==1)
        return;

        vis[row][col]=1;    // marking the current row and column as visited.
        count++;

        dfs(row-1,col,grid,vis,count);
        dfs(row,col+1,grid,vis,count);
        dfs(row+1,col,grid,vis,count);
        dfs(row,col-1,grid,vis,count);
    }

    int maxAreaOfIsland(vector<vector<int>>& grid) 
    {
        n=grid.size();
        m=grid[0].size();
        vector<vector<int>> vis(n,vector<int>(m,0));
        int maxCount=0;
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                int count=0;
                if(grid[i][j]==1 and vis[i][j]==0)
                {
                    dfs(i,j,grid,vis,count);
                    maxCount=max(maxCount,count);
                }
            }
        }
        return maxCount;
    }
};

File: 695-max-area-of-island/max-area-of-island.java
class Solution {
    int n, m;
    public int maxAreaOfIsland(int[][] grid)
    {
        n = grid.length;
        m = grid[0].length;
        boolean[][] visited = new boolean[n][m];
        
        int maxArea = 0;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == 1 && !visited[i][j]){
                    maxArea = Math.max(maxArea, dfs(i, j, grid, visited));
                }
            }
        }
        return maxArea;
    }
    
    private int dfs(int row, int col, int[][] grid, boolean[][] visited){
        if(row < 0 || row >= n || col < 0 || col >= m || grid[row][col] == 0 || visited[row][col]){
            return 0;
        }
        
        visited[row][col] = true;
        int count = 1;
        count += dfs(row + 1, col, grid, visited);
        count += dfs(row - 1, col, grid, visited);
        count += dfs(row, col + 1, grid, visited);
        count += dfs(row, col - 1, grid, visited);
        
        return count;
    }
}